File: app/src/main/java/com/example/firefighters/viewmodels/WaterPointViewModel.java
package com.example.firefighters.viewmodels;

import android.app.Activity;

import com.example.firefighters.models.WaterPointModel;
import com.example.firefighters.models.WaterPointModel;
import com.example.firefighters.repositories.WaterPointRepository;
import com.example.firefighters.repositories.WaterPointRepository;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

public class WaterPointViewModel extends ViewModel {
    private WaterPointRepository repository;

    public void init() {
        if (repository != null)
            return;
        repository = WaterPointRepository.getInstance();
    }
    public LiveData<QuerySnapshot> getWaterPointsQuery(DocumentSnapshot lastDocument, int limitCount) {
        return repository.getWaterPointsQuery(lastDocument, limitCount);
    }
    public LiveData<QuerySnapshot> getWaterPointsQuerySnapshot() {
        return repository.getWaterPointsQuerySnapshot();
    }

    public LiveData<Integer> saveWaterPoint(WaterPointModel waterPointModel ) {
        return repository.saveWaterPoint(waterPointModel);
    }
    public LiveData<Integer> deleteWaterPoint(WaterPointModel waterPointModel ) {
        return repository.deleteWaterPoint(waterPointModel);
    }
    public LiveData<Integer> updateWaterPoint(WaterPointModel waterPointModel ) {
        return repository.updateWaterPoint(waterPointModel);
    }

}

File: app/src/main/java/com/example/firefighters/viewmodels/models/WaterPointViewModel.kt
package com.example.firefighters.viewmodels.models

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.firefighters.models.WaterPointModel
import com.example.firefighters.repositories.WaterPointRepository
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.QuerySnapshot

class WaterPointViewModel(private val mRepository: WaterPointRepository) : ViewModel() {

    val allWaterPointsQuery: LiveData<QuerySnapshot>
        get() = mRepository.allWaterPointsQuery

    fun getWaterPointsQuery(
        lastDocument: DocumentSnapshot?,
        limitCount: Int
    ): LiveData<QuerySnapshot> {
        return mRepository.getWaterPointsQuery(lastDocument, limitCount)
    }

    val waterPointsQuerySnapshot: LiveData<QuerySnapshot?>
        get() = mRepository.waterPointsQuerySnapshot

    fun getWaterPointModel(id: Int): LiveData<WaterPointModel?> {
        return mRepository.getWaterPointModel(id)
    }

    fun saveWaterPoint(waterPointModel: WaterPointModel?): LiveData<Int> {
        return mRepository.saveWaterPoint(waterPointModel!!)
    }

    fun deleteWaterPoint(waterPointModel: WaterPointModel?): LiveData<Int> {
        return mRepository.deleteWaterPoint(waterPointModel!!)
    }

    fun updateWaterPoint(waterPointModel: WaterPointModel?): LiveData<Int> {
        return mRepository.updateWaterPoint(waterPointModel!!)
    }

    class Factory(
        private val repository: WaterPointRepository
    ) : ViewModelProvider.NewInstanceFactory() {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return WaterPointViewModel(repository) as T
        }
    }
}
